pulp_rpm/app/downloaders.py

```diff
@@ -94,3 +104,88 @@ async def _run(self, extra_data=None):
         if self._close_session_on_finalize:
             self.session.close()
         return to_return
+
+
+class UlnDownloader(RpmDownloader):
+    """
+    Custom Downloader for ULN repositories.
+
+    Args:
+        uln_user (str): Username for authentification in ULN network
+        uln_pwd (str): password for authentification in ULN network
+        server_url (str): ULN server url. If no url is given, use default oracle
+            ULN server url `https://linux-update.oracle.com/`
+
+    Raises:
+        NoUlnCredentials: If no or not valid ULN credentials are given, this Error will be displayed
+    """
+
+    def __init__(self, *args, uln_user=None, uln_pwd=None, server_url=None, **kwargs):
+        """
+        Initialize the downloader for ULN repositories.
+
+        If no server URL is given, use the server url from oracle linux.
+        """
+        self.uln_user = uln_user
+        self.uln_pwd = uln_pwd
+        if server_url:
+            self.server_url = server_url
+        else:  # if no server_url is given, use oracle linux url
+            self.server_url = "https://linux-update.oracle.com/"
+        self.headers = None
+        self.session_key = None
+
+        super().__init__(*args, **kwargs)
+
+    async def _run(self, extra_data=None):
+        """
+        Download, validate, and compute digests on the `url`. This is a coroutine.
+
+        Once per session the coroutine logs into the ULN account using the
+        uln_username and uln_pwd. Ther returned key is used for authentification
+        for all other downloads.
+
+        This method provides the same return object type and documented in
+        :meth:`~pulpcore.plugin.download.BaseDownloader._run`.
+        """
+        if self.sles_auth_token:
+            auth_param = f"?{self.sles_auth_token}"
+            url = urljoin(self.url, auth_param)
+        else:
+            url = self.url
+
+        parsed = urlparse(self.url)
+        if parsed.scheme == "uln":
+            if self.uln_user and self.uln_pwd:
+                # get ULN Session-key
+                SERVER_URL = os.path.join(self.server_url, "rpc/api")
+
+                # if proxy, do some thing... TO DO
+
+                client = ServerProxy(SERVER_URL)
+                if not self.session_key:
+                    self.session_key = await client["auth.login"](self.uln_user, self.uln_pwd)
+                    if len(self.session_key) != 43:
+                        raise NoUlnCredentials()
+                    self.headers = {"X-ULN-API-User-Key": self.session_key}
+                    await client.close()
+                # build request url from input uri
+                channelLabel = parsed.netloc
+                path = parsed.path.lstrip("/")
+                url = os.path.join(self.server_url, "XMLRPC/GET-REQ", channelLabel, path)
+            else:
+                raise NoUlnCredentials()
```

Comment:
Might be easier to raise the exception if the credentials are missing and just continue otherwise (keeps the indentation down :grin: ):
```python
if not(self.uln_user and self.uln_pwd):
  raise NoUlnCredentials()
# get ULN Session-key
SERVER_URL = os.path.join(self.server_url, "rpc/api")
#...
```

Reply:
Done ;)

Reply:
Or we create a separate `UlnRemote` which defines these fields as required, and then we don't even need this check. :thinking:
I find this idea increasingly attractive from a design point of view. However, I am not sure if there are currently *any* Pulp plugins with more than one type of remote. Breaking new ground risks waking sleeping dragons.